### PR TITLE
Add pre/inference/post timing breakdown to predictors and YOLOView HUD

### PR DESCRIPTION
**Pull request by glenn-jocher**
## Per-stage timing breakdown

Mirrors ultralytics/yolo-flutter-app#527 so both platforms report and display the same timing decomposition.

### Changes
- Every predictor stamps the end of the Vision request (`markInferenceEnd()`) and reports **inference / postprocess** times through new public `YOLOResult` fields (`preMs`/`inferenceMs`/`postMs`). On iOS, Vision performs input scaling inside the request, so `preMs` is documented as near-zero and the scaling cost is included in `inferenceMs`.
- `YOLOView` shows the breakdown as a smaller caption line under the existing `FPS - ms` label (`x.x pre · x.x inference · x.x post`), hidden when values are absent — matching the Flutter showcase HUD.
- **Fixes a timing bug**: single-image `predictOnImage` previously measured `speed` *after* drawing annotations, so plotting time was included. Timing is now captured before drawing in all predictors, matching the camera path and the Android implementation.

### Testing
- `xcodebuild -scheme UltralyticsYOLO -destination 'generic/platform=iOS Simulator' build` succeeds
- Additive public API only (new fields default to 0)

The Flutter plugin's iOS bridge picks these fields up in a follow-up to ultralytics/yolo-flutter-app#527 once this ships in the pod.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
This PR improves the iOS YOLO SDK and app with clearer docs 📚, more detailed performance reporting ⏱️, faster semantic segmentation handling 🎨, and small UI/version updates for a smoother developer and user experience.

### 📊 Key Changes
- Added **per-stage timing metrics** to `YOLOResult`, including:
  - `preMs`
  - `inferenceMs`
  - `postMs`
- Updated all major predictors—classification, detection, segmentation, semantic segmentation, pose, and OBB—to:
  - mark when inference ends
  - report a cleaner breakdown of inference vs. postprocessing
  - exclude annotation drawing from timing results for more accurate performance numbers
- Enhanced `YOLOView` and the sample iOS app UI to display a **live timing breakdown** under FPS, such as preprocess / inference / postprocess ⏱️
- Improved `SemanticSegmenter` to support **new semantic model outputs** that provide a direct class map (`[1, H, W]`) instead of only older logits formats
- Optimized semantic mask rendering by replacing per-pixel color writes with a more efficient packed-color pipeline 🚀
- Refreshed both English and Chinese READMEs to make the project description, features, and repository structure clearer and more concise
- Expanded `docs/performance.md` with a more structured benchmarking guide, standardized backend results, and clearer explanations of real-world iPhone performance 📱
- Bumped package/app version from **8.9.4** to **8.9.5**

### 🎯 Purpose & Impact
- **More accurate performance insight** for developers: timing now separates model execution from postprocessing, making profiling and optimization much easier
- **Better transparency in the app UI**: users and testers can now see where time is spent, not just total FPS
- **Improved semantic segmentation support**: newer exported models with in-graph ArgMax can be used directly, reducing CPU-side work and helping produce sharper masks 🎯
- **Faster semantic postprocessing** may improve responsiveness, especially for high-resolution semantic tasks
- **Cleaner documentation** makes it easier for new users to understand what the repo includes and how to integrate the Swift package
- **Low risk for end users**: most changes are additive or quality-focused, with the biggest practical benefits going to developers benchmarking, debugging, or integrating YOLO26 on iOS

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/yolo-ios-app` 🚀 PR! This is an automated message to help with review readiness—an engineer will assist shortly. Please review the checklist below for a smooth integration ✅

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/yolo-ios-app/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/yolo-ios-app` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: Sources/UltralyticsYOLO/Segmenter.swift [hunk 1]
```diff
@@ -77,6 +79,7 @@ public final class Segmenter: BasePredictor, @unchecked Sendable {
       emptyResult.speed = finishTiming(notify: false)
       return emptyResult
     }
+    markInferenceEnd()
```
Comment (UltralyticsAssistant) [inline Sources/UltralyticsYOLO/Segmenter.swift:80]: 💡 **MEDIUM**: In the synchronous image path, `parseSegmentationRequest(request)` still runs before `markInferenceEnd()`. That means part of segmentation postprocessing is being counted inside `inferenceMs`, while the realtime path marks the boundary before parsing. Move the inference-end mark to immediately after `perform(...)` succeeds so the breakdown is consistent across segmentation code paths.

File: Sources/UltralyticsYOLO/BasePredictor.swift [hunk 2]
```diff
@@ -378,6 +379,26 @@ public class BasePredictor: Predictor, @unchecked Sendable {
...
   }
 
+  /// Marks the end of inference (start of postprocessing). Call immediately when the Vision request completes.
+  func markInferenceEnd() {
+    tInferEnd = CACurrentMediaTime()
+  }
+
+  /// Per-stage timing in milliseconds: (pre, inference, post). On iOS Vision fuses input scaling into the
+  /// request, so `pre` is folded into `inference` and reported as zero. Call after `finishTiming`/`updateTime`.
+  func timingBreakdownMs() -> (pre: Double, inference: Double, post: Double) {
```
Comment (UltralyticsAssistant) [inline Sources/UltralyticsYOLO/BasePredictor.swift:391]: 📝 **LOW**: These stage values are raw per-frame timings, but the realtime predictors still publish `speed` from the EMA-smoothed `t2`. Since `YOLOView` shows the smoothed total on the first line and this raw breakdown on the second, the numbers will often not add up for the same frame. Consider either smoothing the breakdown with the same EMA or attaching the raw total for the same frame when populating realtime `YOLOResult`s.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Nice, focused change overall—the per-stage timing plumbing is consistent across most predictors and the HUD addition is clean. The main thing to fix is the segmentation single-image path, which currently marks inference end too late and misattributes some postprocessing time. Also worth addressing the raw-vs-smoothed timing mismatch so the two HUD lines don't disagree.

💬 Posted 2 inline comments

File: Sources/UltralyticsYOLO/Segmenter.swift (hunk 1 above)
Comment (glenn-jocher) [reply Sources/UltralyticsYOLO/Segmenter.swift:80]: Fixed in c43e586: the guard is split so `markInferenceEnd()` runs right after `perform` and the parse cost lands in postprocess.

File: Sources/UltralyticsYOLO/BasePredictor.swift (hunk 2 above)
Comment (glenn-jocher) [reply Sources/UltralyticsYOLO/BasePredictor.swift:391]: Fixed in c43e586: the camera path now smooths the stage split with the same EMA as `speed` (`applyTimingBreakdown(smoothed: true)`); single-image predicts keep raw per-frame values for benchmarking.

**Comment by UltralyticsAssistant** [comment]
🎉 Excellent merge, @glenn-jocher — this is a strong quality-of-life upgrade for the iOS YOLO SDK and app.

As Leonardo da Vinci said, *“Simplicity is the ultimate sophistication.”* This PR reflects that well: clearer docs, cleaner timing visibility, faster semantic segmentation handling, and thoughtful UI/version polish all make YOLO26 on iOS easier to understand, benchmark, and trust in real-world use.

The added per-stage timing breakdown and semantic segmentation improvements are especially valuable for developers profiling performance and integrating modern models with confidence. Thanks for continuing to raise the bar for clarity, speed, and usability across the Ultralytics ecosystem.
